Refresh ASLR in Security Features (#64)
canonical/ubuntu-security-documentation: docs/security-features/process-memory/aslr.rst

Address Space Layout Randomization (ASLR)
#########################################

Address Space Layout Randomization (ASLR) is a security feature that randomizes
the location of memory allocations, implemented by the kernel and the ELF

> loader. This randomization makes it more difficult for attackers to reliably predict memory

Reviewer: ```suggestion
Address Space Layout Randomization (ASLR) is a security feature that randomizes the location of key memory areas. This is implemented jointly; the kernel randomizes the initial process layout (i.e., the stack and heap), while the user-space Executable and Linkable Format (ELF) loader randomizes the locations of the executable and its shared libraries. This randomization makes it more difficult for attackers to reliably predict memory
```

> addresses when attempting a memory-corruption exploit.

Reviewer 2: What about reading memory?

> addresses when attempting a memory-corruption exploit.

Reviewer: ```suggestion
loader. This randomization makes it more difficult for attackers to predict memory addresses, which helps prevent both unauthorized memory reads (information leaks) and memory-corruption exploits.
```

by default.

Configuration
=============

ASLR is controlled system-wide by the value of ``/proc/sys/kernel/randomize_va_space``.

.. list-table:: ``/proc/sys/kernel/randomize_va_space`` values
   :widths: 15 85
   :header-rows: 1

   * - Value
     - Meaning
   * - 0
     - ASLR is off.
   * - 1
     - Partial ASLR. The addresses of stack, mmap base and vDSO page are randomized.
       This is the default if the ``CONFIG_COMPAT_BRK`` option is enabled.
   * - 2 (default for most systems)
     - Full ASLR. Additionally enable heap randomization.
       This is the default if ``CONFIG_COMPAT_BRK`` is disabled.

To temporarily change the value, it can be done via the ``proc`` file system::

   echo 1 > /proc/sys/kernel/randomize_va_space

Or via ``sysctl``::

   sysctl -w kernel.randomize_va_space=1

To make it persistent across reboots, add a line to a config file in ``/etc/sysctl.d/``::

   kernel.randomize_va_space=1

.. _types-of-aslr:

Types of ASLR
=============

See `test-kernel-security.py <https://git.launchpad.net/qa-regression-testing/tree/scripts/test-kernel-security.py>`__ for regression tests for all the different types of ASLR.

.. _stack-aslr:

Stack ASLR
~~~~~~~~~~

> *Introduced in the Linux kernel:* 2.6.15 (Ubuntu 6.06)

Reviewer: The style guide wants to ensure that all releases follow the format `Ubuntu XX.XX [LTS] (Code Name)`. In this case it would be `Ubuntu 6.06 (Dapper Drake)`. But given the age of the reference I would opt to drop this reference entirely as I don't think the historical reference is needed. If you do go that route, just be sure to delete the other instances that reference the old release as well :)

> Each execution of a program results in a different stack memory space layout.

Reviewer 2: Layout or just the starting address of the stack?

> This makes it harder to locate in memory where to attack or deliver an

Reviewer 2: This doesn't read right.

> executable attack payload.

Reviewer: Maybe something like this?
```suggestion
This makes it harder to locate where to attack or deliver an executable payload in memory.
```

.. _mmap-aslr:

Libs/mmap ASLR
~~~~~~~~~~~~~~

*Introduced in the Linux kernel:* 2.6.15 (Ubuntu 6.06 LTS)

> Each execution of a program results in a different ``mmap`` memory space layout.

Reviewer 2: This is fairly confusing. Isn't there just a single random base that is then used as an offset for subsequent mmap locations (i.e. libraries will still have the same relative offsets)?

Reviewer: ```suggestion
Each execution of a program results in a different base address for the ``mmap`` memory region, changing the absolute memory locations of mapped files and libraries.
```

Author: Fixed, thanks!

> It causes the dynamically loaded libraries to get loaded into different

Reviewer: Agreed, I would be explicit and mention that ASLR applies to the stack, heap, libraries, and memory-mapped regions.

> It causes the dynamically loaded libraries to get loaded into different

Author: Changed a bit, but I left out the stack part as the stack pointer is not related to `mmap`?

locations each time. This makes it harder to locate in memory where to jump to
for "return-to-libc" to similar attacks.

.. _exec-aslr:

Exec ASLR
~~~~~~~~~

*Introduced in the Linux kernel:* 2.6.25, backported to Ubuntu 8.04 LTS

Each execution of a program that has been built with "-fPIE -pie"
(see :ref:`Built as PIE <compiler-flags-pie>`) will get loaded into a different memory location.

This makes it harder to locate in memory where to attack or jump to when
performing memory-corruption-based attacks.

.. _brk-aslr:

``brk`` ASLR
~~~~~~~~~~~~

*Introduced in the Linux kernel:* 2.6.26 (Ubuntu 8.10)

> Small ``malloc`` allocations are served from the program break (``brk``)

Reviewer 2: You should mention this is the heap, as it is a far more common term. See man 2 brk and man 3 malloc. Might also sound nicer to just say the location where the executable (both code and static data) is loaded in memory and the heap.

Author: Fixed, thanks!

segment. Randomizing the gap between the ``exec`` region and ``brk`` makes it
harder to locate in memory where to attack or jump to.

.. _vdso-aslr:

vDSO ASLR
~~~~~~~~~

*Introduced in the Linux kernel:* 2.6.18 (x86/PPC), 2.6.22 (x86_64) (Ubuntu 8.04 LTS)

Each execution of a program results in a random vDSO location.

The vDSO (Virtual Dynamic Shared Object) offers a selected set of kernel space

> routines (e.g. ``gettimeofday``) to user space applications to improve

Reviewer: So maybe something like `... improve the performance of system calls`? Also I think the avoidance of "jump-into-syscall" attacks covers your second point, but maybe that could be more clear in case the audience doesn't know about these attacks. 

I think either linking to a reference for "jump-into-syscalls" attacks (really didn't find anything good except https://www.paloaltonetworks.com/blog/security-operations/a-deep-dive-into-malicious-direct-syscall-detection/ and I don't think we want to link to another companies resources) or changing this to be something like `This performance gain is possible thanks to vDSO allowing what are normally syscalls to be executed entirely in user space, without jumping to kernel space.` would address this.
